# Aufgabe 7: AutoML

AutoML stellt Methoden und Prozesse zur Verfügung, um maschinelles Lernen auch Nicht-Machine-Learning-Experten zugänglich zu machen. Die Universität Freiburg hat eine Internetseite mit weiterführenden Informationen und einigen AutoML-Packages zusammengestellt, die unter folgenden Link zu erreichen ist: https://www.ml4aad.org/automl/.

###### (a) Wählen Sie ein AutoML Package. Begründen Sie Ihre Auswahl.

Wir haben uns für _H2O AutoML_ entschieden, weil es in Hinblick auf die Reproduzierbarkeit der Ergebnisse am leichtesten auf verschiedenen Plattformen installierbar ist.

Bei _auto-sklearn_ beispielsweise fehlten die Berechtigungen, um die benötigten Dependencies auf SDIL zu installieren.
Für andere AutoML-Bibliotheken wurde lediglich ein git-repo zur Verfügung gestellt, weshalb diese nicht einfach über eine `requirements.txt` installiert werden können.

Zudem fanden wir die Dokumentation zu _H2O AutoML_ auf den ersten Blick sehr übersichtlich.

###### (b) Führen Sie die Klassifikationsaufgabe von Aufgabe 3 mit AutoML durch. Vergleichen Sie die Ergebnisse mit den Ergebnissen aus Aufgabe 3.

http://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html

In [1]:
# Import Libraries
import time
import multiprocessing

import pandas as pd
import numpy as np

from sklearn import svm
from sklearn import model_selection
from sklearn.model_selection import GridSearchCV

import h2o
from h2o.automl import H2OAutoML
from h2o.frame import H2OFrame

from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

# Fix for display error of tables
from IPython.core.display import HTML
HTML("<style>.rendered_html table {width:auto !important;}</style>")

In [2]:
# initialize h2o
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "1.8.0_152-release"; OpenJDK Runtime Environment (build 1.8.0_152-release-1056-b12); OpenJDK 64-Bit Server VM (build 25.152-b12, mixed mode)
  Starting server from /Users/d062356/miniconda3/envs/smart-data-analytics/h2o_jar/h2o.jar
  Ice root: /var/folders/g3/8_fmjvcd5t3fqv5m5c3zz20m0000gn/T/tmprqtgkwgs
  JVM stdout: /var/folders/g3/8_fmjvcd5t3fqv5m5c3zz20m0000gn/T/tmprqtgkwgs/h2o_d062356_started_from_python.out
  JVM stderr: /var/folders/g3/8_fmjvcd5t3fqv5m5c3zz20m0000gn/T/tmprqtgkwgs/h2o_d062356_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321... successful.


--------------------------  ----------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       Europe/Berlin
H2O data parsing timezone:  UTC
H2O cluster version:        3.18.0.2
H2O cluster version age:    1 year, 2 months and 19 days !!!
H2O cluster name:           H2O_from_python_d062356_5ql9sa
H2O cluster total nodes:    1
H2O cluster free memory:    3.556 Gb
H2O cluster total cores:    8
H2O cluster allowed cores:  8
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.8 final
--------------------------  ----------------------------------------

In [3]:
# Load dataset iris
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']
iris = pd.read_csv(url, names=names)

In [4]:
# Split-out validation dataset
array = iris.values
X = array[:,0:4]
Y = array[:,4]
test_size = 0.20
seed = 7
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, Y, test_size=test_size, random_state=seed)

In [5]:
# prepare data by creating a train and test array and transforming it into an h2o frame
train_array = np.column_stack([X_train, y_train])
test_array = np.column_stack([X_test, y_test])

train = H2OFrame(python_obj=train_array, column_names=names)
test = H2OFrame(python_obj=test_array, column_names=names)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [6]:
# Identify predictors and response
x = names
y = "class"
x.remove(y)

In [7]:
# set number of models to 20 and start training
aml = H2OAutoML(max_models=20, seed=1, nfolds=10)
aml.train(x=x, y=y, training_frame=train)                                 

AutoML progress: |████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [8]:
# View the AutoML Leaderboard
lb = aml.leaderboard
display(lb.head(rows=lb.nrows)) # Print all rows instead of default (10 rows)

model_id mean_per_class_error 
 
 
 GBM_grid_0_AutoML_20190525_162151_model_15 0.011111 
 DRF_0_AutoML_20190525_162151 0.011111 
 XRT_0_AutoML_20190525_162151 0.011111 
 DeepLearning_grid_0_AutoML_20190525_162151_model_0 0.011111 
 GBM_grid_0_AutoML_20190525_162151_model_8 0.022222 
 GBM_grid_0_AutoML_20190525_162151_model_11 0.022222 
 GLM_grid_0_AutoML_20190525_162151_model_0 0.022222 
 GBM_grid_0_AutoML_20190525_162151_model_9 0.022222 
 StackedEnsemble_AllModels_0_AutoML_20190525_162151 0.022222 
 GBM_grid_0_AutoML_20190525_162151_model_2 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_14 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_12 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_10 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_1 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_13 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_3 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_5 0.033333 
 GBM_grid_0_AutoML_20190525_162151_model_0 0.033333 
 DeepLearning_0_AutoML_20190525_162151 0.044444 
 GBM_grid_0_AutoML_20190525_162151_model_7 0.144444 
 GBM_grid_0_AutoML_20190525_162151_model_4 0.144444 
 StackedEnsemble_BestOfFamily_0_AutoML_20190525_162151 0.666667

In [9]:
# The leader model is stored here
display(aml.leader)

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_grid_0_AutoML_20190525_162151_model_15


ModelMetricsMultinomial: gbm
** Reported on train data. **

MSE: 8.349901175356046e-06
RMSE: 0.002889619555470243
LogLoss: 0.0007618680879659699
Mean Per-Class Error: 0.0
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error    Rate
-------------  -----------------  ----------------  -------  ------
29             0                  0                 0        0 / 29
0              30                 0                 0        0 / 30
0              0                  30                0        0 / 30
29             30                 30                0        0 / 89

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    1
2    1
3    1


ModelMetricsMultinomial: gbm
** Reported on validation data. **

MSE: 0.029858859578529552
RMSE: 0.17279716310903243
LogLoss: 0.10589490037197341
Mean Per-Class Error: 0.037037037037037035
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error      Rate
-------------  -----------------  ----------------  ---------  ------
14             0                  0                 0          0 / 14
0              8                  0                 0          0 / 8
0              1                  8                 0.111111   1 / 9
14             9                  8                 0.0322581  1 / 31

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.967742
2    1
3    1


ModelMetricsMultinomial: gbm
** Reported on cross-validation data. **

MSE: 0.011864144793886748
RMSE: 0.1089226550993261
LogLoss: 0.03798343286985158
Mean Per-Class Error: 0.011111111111111112
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error      Rate
-------------  -----------------  ----------------  ---------  ------
29             0                  0                 0          0 / 29
0              30                 0                 0          0 / 30
0              1                  29                0.0333333  1 / 30
29             31                 29                0.011236   1 / 89

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.988764
2    1
3    1

Cross-Validation Metrics Summary: 


mean        sd         cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid    cv_6_valid    cv_7_valid    cv_8_valid    cv_9_valid    cv_10_valid
-----------------------  ----------  ---------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  -------------
accuracy                 0.988889    0.0235702  1             1             1             1             1             1             1             0.888889      1             1
err                      0.0111111   0.0235702  0             0             0             0             0             0             0             0.111111      0             0
err_count                0.1         0.212132   0             0             0             0             0             0             0             1             0             0
logloss                  0.0375999   0.0299332  0.0370365     4.54959e-05   5.98264e-08   0.0738255     0.0766432     0.0125085     3.29466e-09   0.13271       0.0397612     0.00346907
max_per_class_error      0.0166667   0.0353553  0             0             0             0             0             0             0             0.166667      0             0
mean_per_class_accuracy  0.994444    0.0117851  1             1             1             1             1             1             1             0.944444      1             1
mean_per_class_error     0.00555556  0.0117851  0             0             0             0             0             0             0             0.0555556     0             0
mse                      0.0117332   0.0110742  0.00871758    1.54135e-08   1.46006e-14   0.0252963     0.0251847     0.00117198    3.38481e-17   0.049178      0.00770637    7.67729e-05
r2                       0.975592    0.025343   0.988611      1             1             0.962055      0.92154       0.997842      1             0.895173      0.99082       0.999874
rmse                     0.0763781   0.0543119  0.093368      0.000124151   1.20833e-07   0.159048      0.158697      0.0342342     5.81792e-09   0.221761      0.087786      0.00876202

Scoring History: 


timestamp            duration    number_of_trees    training_rmse         training_logloss       training_classification_error    validation_rmse      validation_logloss    validation_classification_error
---  -------------------  ----------  -----------------  --------------------  ---------------------  -------------------------------  -------------------  --------------------  ---------------------------------
     2019-05-25 16:22:49  37.129 sec  0.0                0.6666666666666661    1.0986122886681084     0.6292134831460674               0.6666666666666667   1.0986122886681104    0.7419354838709677
     2019-05-25 16:22:49  37.132 sec  5.0                0.5506816410787105    0.8005047392174571     0.033707865168539325             0.5457820812084995   0.789730804187998     0.03225806451612903
     2019-05-25 16:22:49  37.135 sec  10.0               0.4380941375026858    0.5758319796326994     0.0                              0.43460344389895766  0.5698681624396363    0.03225806451612903
     2019-05-25 16:22:49  37.138 sec  15.0               0.34115437086893574   0.414513275177708      0.011235955056179775             0.33515789298402787  0.4052221538536994    0.03225806451612903
     2019-05-25 16:22:49  37.141 sec  20.0               0.2762059569961727    0.31761948174665694    0.0                              0.27553773980342994  0.31453070311770664   0.03225806451612903
---  ---                  ---         ---                ---                   ---                    ---                              ---                  ---                   ---
     2019-05-25 16:22:49  37.311 sec  240.0              0.003966178416756533  0.001040356152368516   0.0                              0.172198066257037    0.10327573830849818   0.03225806451612903
     2019-05-25 16:22:49  37.315 sec  245.0              0.003741490382055205  0.0009759689058308045  0.0                              0.17180679976543806  0.10157121078458872   0.03225806451612903
     2019-05-25 16:22:49  37.319 sec  250.0              0.003377749644207863  0.0008805145220108146  0.0                              0.17251256132173615  0.10460434918776602   0.03225806451612903
     2019-05-25 16:22:49  37.323 sec  255.0              0.003010486595760493  0.0007844860623340406  0.0                              0.17322745287647398  0.10799511735948097   0.03225806451612903
     2019-05-25 16:22:49  37.325 sec  256.0              0.002889619555470243  0.0007618680879659699  0.0                              0.17279716310903243  0.10589490037197341   0.03225806451612903


See the whole table with table.as_data_frame()
Variable Importances: 


variable      relative_importance    scaled_importance    percentage
------------  ---------------------  -------------------  ------------
petal-length  249.942                1                    0.487167
petal-width   176.807                0.707391             0.344618
sepal-length  58.3064                0.23328              0.113646
sepal-width   27.9966                0.112013             0.0545689

In [10]:
# predict on test data
predictions = aml.leader.predict(test).as_data_frame()['predict'].tolist()

gbm prediction progress: |████████████████████████████████████████████████| 100%


In [11]:
# get target for test data
y_test = test.as_data_frame()['class'].tolist()

In [12]:
# print results
print(accuracy_score(y_test, predictions))
print(confusion_matrix(y_test, predictions))
print(classification_report(y_test, predictions))

0.8666666666666667
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



Der DeepLearning-Ansatz wird von AutoML als der geeignetste betrachtet. Dieser erreicht auf den Trainingsdaten eine Accuracy von 0.8667 (Deep Learning-Ansätze sind hier leider nicht reproduzierbar: "H2O Deep Learning models are not reproducible by default for performance reasons", http://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html#automl-interface)und liegt damit unterhalb der Accuracy des Models mit der höchsten Accuracy aus Aufgabe drei (0.967). 

# Zweiter Datensatz

In [13]:
# Load dataset
url_heart = "./data/heart.csv"
names_heart =  ['age', 'sex', 'chest_pain_type', 'resting_blood_pressure', 'cholesterol', 
                 'fasting_blood_sugar', 'rest_ecg', 'max_heart_rate_achieved', 'exercise_induced_angina',
                 'st_depression', 'st_slope', 'num_major_vessels', 'thalassemia', 'target']
heart = pd.read_csv(url_heart, names=names_heart)
heart.head()

age  sex chest_pain_type resting_blood_pressure cholesterol  \
0  age  sex              cp               trestbps        chol   
1   63    1               3                    145         233   
2   37    1               2                    130         250   
3   41    0               1                    130         204   
4   56    1               1                    120         236   

  fasting_blood_sugar rest_ecg max_heart_rate_achieved  \
0                 fbs  restecg                 thalach   
1                   1        0                     150   
2                   0        1                     187   
3                   0        0                     172   
4                   0        1                     178   

  exercise_induced_angina st_depression st_slope num_major_vessels  \
0                   exang       oldpeak    slope                ca   
1                       0           2.3        0                 0   
2                       0           3.5        0                 0   
3                       0           1.4        2                 0   
4                       0           0.8        2                 0   

  thalassemia  target  
0        thal  target  
1           1       1  
2           2       1  
3           2       1  
4           2       1

In [14]:
# drop first row
heart = heart.drop(heart.index[0])
heart.head()

age sex chest_pain_type resting_blood_pressure cholesterol  \
1  63   1               3                    145         233   
2  37   1               2                    130         250   
3  41   0               1                    130         204   
4  56   1               1                    120         236   
5  57   0               0                    120         354   

  fasting_blood_sugar rest_ecg max_heart_rate_achieved  \
1                   1        0                     150   
2                   0        1                     187   
3                   0        0                     172   
4                   0        1                     178   
5                   0        1                     163   

  exercise_induced_angina st_depression st_slope num_major_vessels  \
1                       0           2.3        0                 0   
2                       0           3.5        0                 0   
3                       0           1.4        2                 0   
4                       0           0.8        2                 0   
5                       1           0.6        2                 0   

  thalassemia target  
1           1      1  
2           2      1  
3           2      1  
4           2      1  
5           2      1

In [15]:
# Split-out validation dataset
array_heart = heart.values
X_heart = array_heart[:,0:13]
Y_heart = array_heart[:,13]
test_size_heart = 0.20
seed_heart = 10
X_heart_train, X_heart_test, y_heart_train, y_heart_test = model_selection.train_test_split(X_heart, Y_heart, test_size=test_size_heart, random_state=seed_heart)

In [16]:
# prepare data by creating a train and test array and transforming it into an h2o frame
train_heart_array = np.column_stack([X_heart_train, y_heart_train])
test_heart_array = np.column_stack([X_heart_test, y_heart_test])

train_heart = H2OFrame(python_obj=train_heart_array, column_names=names_heart)
test_heart = H2OFrame(python_obj=test_heart_array, column_names=names_heart)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [17]:
# Identify predictors and response
x_heart = names_heart
y_heart = "target"
x_heart.remove(y_heart)

In [18]:
# set number of models to 20 and start training
aml_heart = H2OAutoML(max_models=20, seed=1, nfolds=10)
aml_heart.train(x=x_heart, y=y_heart, training_frame=train_heart)                                 

AutoML progress: |████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [19]:
# View the AutoML Leaderboard
lb_heart = aml_heart.leaderboard
display(lb_heart.head(rows=lb_heart.nrows)) # Print all rows instead of default (10 rows)

model_id mean_residual_deviance rmse mae rmsle 
 
 
 StackedEnsemble_AllModels_0_AutoML_20190525_162343 0.111374 0.333728 0.246481 0.234103 
 StackedEnsemble_BestOfFamily_0_AutoML_20190525_162343 0.112304 0.335118 0.248627 0.234248 
 DeepLearning_grid_0_AutoML_20190525_162343_model_0 0.121459 0.34851 0.261562 0.251839 
 XRT_0_AutoML_20190525_162343 0.123401 0.351284 0.253637 0.245572 
 GLM_grid_0_AutoML_20190525_162343_model_0 0.12545 0.35419 0.284097 0.24988 
 DeepLearning_grid_0_AutoML_20190525_162343_model_5 0.130154 0.360768 0.274528 0.257425 
 GBM_grid_0_AutoML_20190525_162343_model_5 0.130494 0.36124 0.29209 0.253636 
 DeepLearning_grid_0_AutoML_20190525_162343_model_2 0.131651 0.362837 0.286262 0.256 
 DRF_0_AutoML_20190525_162343 0.131691 0.362893 0.249952 0.251953 
 GBM_grid_0_AutoML_20190525_162343_model_2 0.133861 0.365871 0.270722 0.254411 
 GBM_grid_0_AutoML_20190525_162343_model_3 0.134076 0.366163 0.270683 0.254664 
 GBM_grid_0_AutoML_20190525_162343_model_1 0.1343 0.36647 0.271343 0.254619 
 DeepLearning_grid_0_AutoML_20190525_162343_model_6 0.13618 0.369026 0.27593 0.263038 
 DeepLearning_grid_0_AutoML_20190525_162343_model_4 0.136224 0.369086 0.280046 0.277382 
 DeepLearning_grid_0_AutoML_20190525_162343_model_1 0.139071 0.372922 0.287504 0.26295 
 GBM_grid_0_AutoML_20190525_162343_model_9 0.141071 0.375594 0.269175 0.260894 
 GBM_grid_0_AutoML_20190525_162343_model_8 0.142745 0.377817 0.263601 0.261166 
 DeepLearning_grid_0_AutoML_20190525_162343_model_3 0.151023 0.388617 0.300207 0.284195 
 GBM_grid_0_AutoML_20190525_162343_model_0 0.151278 0.388944 0.275203 0.269308 
 DeepLearning_0_AutoML_20190525_162343 0.152558 0.390587 0.304398 0.2869 
 GBM_grid_0_AutoML_20190525_162343_model_7 0.244261 0.494228 0.490204 0.347241 
 GBM_grid_0_AutoML_20190525_162343_model_4 0.245322 0.4953 0.491239 0.34795

In [20]:
# The leader model is stored here
display(aml_heart.leader)

Model Details
H2OStackedEnsembleEstimator :  Stacked Ensemble
Model Key:  StackedEnsemble_AllModels_0_AutoML_20190525_162343
No model summary for this model


ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 0.003931911832168075
RMSE: 0.06270495859314536
MAE: 0.04907276895578059
RMSLE: 0.0446905019968733
R^2: 0.9840227074756345
Mean Residual Deviance: 0.003931911832168075
Null degrees of freedom: 191
Residual degrees of freedom: 185
Null deviance: 47.25
Residual deviance: 0.7549270717762703
AIC: -502.54446726863966

ModelMetricsRegressionGLM: stackedensemble
** Reported on validation data. **

MSE: 0.12618346556015075
RMSE: 0.35522312081303314
MAE: 0.2586696685420625
RMSLE: 0.23477350712671397
R^2: 0.46441652988051463
Mean Residual Deviance: 0.12618346556015075
Null degrees of freedom: 49
Residual degrees of freedom: 43
Null deviance: 11.9453125
Residual deviance: 6.309173278007538
AIC: 54.392935560116854

ModelMetricsRegressionGLM: stackedensemble
** Repo

In [21]:
# predict on test data
predictions_heart = aml_heart.leader.predict(test_heart).as_data_frame()['predict'].tolist()

stackedensemble prediction progress: |████████████████████████████████████| 100%


In [22]:
# get target for test data
y_heart_test = test_heart.as_data_frame()['target'].tolist()

# predictions are continuous --> set threshold for final prediction to 0.5
predictions_heart = pd.DataFrame(predictions_heart)
mask = predictions_heart>0.5
predictions_heart_binary = mask.astype(int)

In [23]:
# for comparability with ex 3
def sensitivity_score(y_true, y_pred):
    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
    sensitivity = tp / (tp + fn)
    return sensitivity

def specificity_score(y_true, y_pred):
    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
    specificity = tn / (tn + fp)
    return specificity

In [24]:
# print results
print('accuracy: '+ str(accuracy_score(y_heart_test, predictions_heart_binary)))
print('sensitivity:' + str(sensitivity_score(y_heart_test, predictions_heart_binary)))
print('specificity:' + str(specificity_score(y_heart_test, predictions_heart_binary)))

#print(confusion_matrix(y_heart_test, predictions_heart_binary))
#print(classification_report(y_heart_test, predictions_heart_binary))

accuracy: 0.8032786885245902
sensitivity:0.8461538461538461
specificity:0.7714285714285715


Hier ist die Accuracy mit ca. 0,8033 identisch zu der in Aufgabe 3 erreichten. Sensitivity weicht mit 0,8462 relativ gering von der in Aufgabe 3 erreichten (0,8077) ab. Auch die Specificity weicht mit 0,0,7714 von der in Aufgabe 3 erreichten (0,8). Ohne dies mit einem statistischen Test zu belegen, werden die Ergebnisse als mit denen der Aufgabe drei vergleichbar eingeschätzt.

###### (c) Was ist Ihre Meinung zu AutoML?

- AutoML nimmt einem viel Arbeit ab, was gerade für Anfänger hilfreich ist. Allerdings verliert man hierdurch auch etwas Kontrolle. Die Ergebnisse sind scheinbar schlechter als in Aufgabe 3. Jedoch ist es schwer nachzuvollziehen wieso, da eben alles automatisch passiert. Zudem ist nicht klar, wieso welche Modelle getestet wurden. 

- "AutoML can only guarantee reproducibility under certain conditions. H2O Deep Learning models are not reproducible by default for performance reasons", dies ist natürlich sehr ungünstig.http://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html#automl-interface, die eingeschränkte Reproduzierbarkeit ist ein Problem.

- Das Training dauert relativ lange, weshalb es etwas enttäuschend ist, das beim Iris-Datensatz keine zu Aufgabe 3 vergleichbaren Ergebnisse erzielt werden konnten. 

- "goal of finding the “best” model without any prior knowledge", https://medium.com/analytics-vidhya/gentle-introduction-to-automl-from-h2o-ai-a42b393b4ba2 --> diese Erfahrung können wir zumindest hinsichtlich des Datensatzes Iris nicht teilen. Man kann mit wenig Ahnung ein akzeptables Ergebnis erreichen, jedoch waren unsere Ergebnisse nicht im sehr guten Bereich.

- "The current version of AutoML trains and cross-validates the following algorithms (in the following order): three pre-specified XGBoost GBM (Gradient Boosting Machine) models, a fixed grid of GLMs, a default Random Forest (DRF), five pre-specified H2O GBMs, a near-default Deep Neural Net, an Extremely Randomized Forest (XRT), a random grid of XGBoost GBMs, a random grid of H2O GBMs, and a random grid of Deep Neural Nets. In some cases, there will not be enough time to complete all the algorithms, so some may be missing from teh leaderboard. AutoML then trains two Stacked Ensemble models." http://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html --> Unseres Erachtens nach sollten noch mehr verschiedene Modelle ausprobiert werden. 